File: scripts/migrate.mjs
import fs from 'node:fs/promises';
import path from 'node:path';

const SRC_DIR = path.join(process.cwd(), 'allfiles', 'c_Programs');
const DEST_DIR = path.join(process.cwd(), 'src', 'content', 'docs', 'c-programming');

async function migrate() {
    console.log('Starting migration...');

    try {
        const categories = await fs.readdir(SRC_DIR);

        for (const category of categories) {
            if (category.endsWith('.md')) continue; // Skip raw markdowns

            const categoryPath = path.join(SRC_DIR, category);
            const stat = await fs.stat(categoryPath);

            if (!stat.isDirectory()) continue;

            const files = await fs.readdir(categoryPath);

            // Create dest directory first
            const destCategoryPath = path.join(DEST_DIR, category.toLowerCase());
            await fs.mkdir(destCategoryPath, { recursive: true });

            for (const file of files) {
                if (!file.endsWith('.c')) continue;

                const filePath = path.join(categoryPath, file);
                const fileContent = await fs.readFile(filePath, 'utf-8');

                const title = file.replace('.c', '').replace(/_/g, ' ');
                const slug = file.replace('.c', '').replace(/_/g, '-').toLowerCase();

                let difficulty = 'Beginner';
                if (category.includes('Array') || category.includes('Function') || category.includes('Pointer')) {
                    difficulty = 'Intermediate';
                }

                let tags = ['C', category.replace(/_/g, ' ')];

                const mdxContent = `---
title: "${title}"
description: "C Program for ${title}"
difficulty: "${difficulty}"
tags: ${JSON.stringify(tags)}
---

import DifficultyBadge from '../../../../components/DifficultyBadge.astro';

<DifficultyBadge level={frontmatter.difficulty} />

### Code Example

\`\`\`c
${fileContent}
\`\`\`
`;

                const destFilePath = path.join(destCategoryPath, `${slug}.mdx`);
                await fs.writeFile(destFilePath, mdxContent);
                console.log(`Migrated: ${file} -> ${destCategoryPath}/${slug}.mdx`);
            }
        }

        console.log('Migration Complete!');
    } catch (err) {
        console.error('Migration failed:', err);
    }
}

migrate();

File: scripts/migrate.py
import os
import shutil
import json

src_dir = os.path.join(os.getcwd(), 'allfiles', 'c_Programs')
dest_dir = os.path.join(os.getcwd(), 'src', 'content', 'docs', 'c-programming')

def migrate():
    print('Starting migration...')
    
    try:
        categories = os.listdir(src_dir)
        
        for category in categories:
            if category.endswith('.md'):
                continue
            
            category_path = os.path.join(src_dir, category)
            if not os.path.isdir(category_path):
                continue
                
            files = os.listdir(category_path)
            
            dest_category_path = os.path.join(dest_dir, category.lower())
            os.makedirs(dest_category_path, exist_ok=True)
            
            for file in files:
                if not file.endswith('.c'):
                    continue
                    
                file_path = os.path.join(category_path, file)
                
                with open(file_path, 'r', encoding='utf-8') as f:
                    file_content = f.read()
                
                title = file.replace('.c', '').replace('_', ' ')
                slug = file.replace('.c', '').replace('_', '-').lower()
                
                difficulty = 'Beginner'
                if 'Array' in category or 'Function' in category or 'Pointer' in category:
                    difficulty = 'Intermediate'
                
                tags = ['C', category.replace('_', ' ')]
                tags_json = json.dumps(tags)
                
                mdx_content = f"""---
title: "{title}"
description: "C Program for {title}"
difficulty: "{difficulty}"
tags: {tags_json}
---

import DifficultyBadge from '../../../../components/DifficultyBadge.astro';

<DifficultyBadge level="{difficulty}" />

### Code Example

```c
{file_content}
```
"""
                
                dest_file_path = os.path.join(dest_category_path, f"{slug}.mdx")
                with open(dest_file_path, 'w', encoding='utf-8') as f:
                    f.write(mdx_content)
                print(f"Migrated: {file} -> {dest_category_path}/{slug}.mdx")
                
        print('Migration Complete!')
    except Exception as e:
        print(f'Migration failed: {e}')

if __name__ == '__main__':
    migrate()
